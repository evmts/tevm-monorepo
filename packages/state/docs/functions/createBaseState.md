[**@tevm/state**](../README.md) • **Docs**

***

[@tevm/state](../globals.md) / createBaseState

# Function: createBaseState()

`Internal`

> **createBaseState**(`options`): [`BaseState`](../type-aliases/BaseState.md)

Creates the core data structure for ethereum state

## Parameters

• **options**: [`StateOptions`](../type-aliases/StateOptions.md)

## Returns

[`BaseState`](../type-aliases/BaseState.md)

Creates the core data structure the state manager operates on

## Source

[packages/state/src/createBaseState.js:20](https://github.com/evmts/tevm-monorepo/blob/main/packages/state/src/createBaseState.js#L20)
