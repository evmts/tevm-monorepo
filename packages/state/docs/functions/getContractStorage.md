[**@tevm/state**](../README.md) • **Docs**

***

[@tevm/state](../globals.md) / getContractStorage

# Function: getContractStorage()

> **getContractStorage**(`baseState`): (`address`, `key`) => `Promise`\<`Uint8Array`\>

Gets the storage value associated with the provided `address` and `key`. This method returns
the shortest representation of the stored value.
corresponding to the provided address at the provided key.
If this does not exist an empty `Uint8Array` is returned.

## Parameters

• **baseState**: [`BaseState`](../type-aliases/BaseState.md)

## Returns

`Function`

### Parameters

• **address**: `Address`

• **key**: `Uint8Array`

### Returns

`Promise`\<`Uint8Array`\>

## Source

[packages/state/src/actions/getContractStorage.js:13](https://github.com/evmts/tevm-monorepo/blob/main/packages/state/src/actions/getContractStorage.js#L13)
