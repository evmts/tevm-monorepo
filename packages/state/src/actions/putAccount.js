/**
 * Saves an account into state under the provided `address`.
 * @type {import("../state-types/index.js").StateAction<'putAccount'>}
 */
export const putAccount = (baseState) => async (address, account) => {
	if (account !== undefined) {
		baseState.caches.accounts?.put(address, account)
	} else {
		baseState.caches.accounts?.del(address)
	}
}
