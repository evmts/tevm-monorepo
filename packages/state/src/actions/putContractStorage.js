import { getAccount } from './getAccount.js'

/**
 * @param {Uint8Array} bytes
 * @returns {Uint8Array}
 */
const stripZeros = (bytes) => {
	if (!(bytes instanceof Uint8Array)) {
		throw new Error('Unexpected type')
	}
	return bytes.slice(bytes.findIndex(/** @param {number} entry*/ (entry) => entry !== 0))
}
/**
 * Adds value to the cache for the `account`
 * corresponding to `address` at the provided `key`.
 * Cannot be more than 32 bytes. Leading zeros are stripped.
 * If it is empty or filled with zeros, deletes the value.
 * @type {import("../state-types/index.js").StateAction<'putContractStorage'>}
 */
export const putContractStorage = (baseState) => async (address, key, value) => {
	if (key.length !== 32) {
		throw new Error('Storage key must be 32 bytes long!')
	}

	const account = await getAccount(baseState)(address)
	if (!account) {
		throw new Error('cannot putContractStorage on non existing acccount!')
	}
	baseState.caches.storage.put(address, key, stripZeros(value))
}
