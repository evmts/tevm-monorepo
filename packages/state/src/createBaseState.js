import { AccountCache, CacheType, StorageCache } from '@ethereumjs/statemanager'
import { createLogger } from '@tevm/logger'
import { ContractCache } from './ContractCache.js'
import { checkpoint } from './actions/checkpoint.js'
import { commit } from './actions/commit.js'
import { generateCanonicalGenesis } from './actions/generateCannonicalGenesis.js'

/**
 * @type {import('viem').Hex}
 */
const INITIAL_STATE_ROOT = '0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421'

/**
 * @internal
 * Creates the core data structure for ethereum state
 * @param {import('./state-types/index.js').StateOptions} options
 * @returns {import('./BaseState.js').BaseState}
 * Creates the core data structure the state manager operates on
 */
export const createBaseState = (options) => {
	const logger = createLogger({
		level: options.loggingLevel ?? 'warn',
		name: '@tevm/state-manager',
	})
	/**
	 * @type {import('./state-types/StateRoots.js').StateRoots}
	 */
	const stateRoots = options.stateRoots ?? new Map()
	if (options.genesisState && options.currentStateRoot) {
		stateRoots.set(options.currentStateRoot, options.genesisState)
	} else {
		stateRoots.set(INITIAL_STATE_ROOT, options.genesisState ?? {})
	}
	let currentStateRoot = options.currentStateRoot ?? INITIAL_STATE_ROOT
	/**
	 * @type {import('./BaseState.js').BaseState}
	 */
	const state = {
		logger,
		getCurrentStateRoot: () => currentStateRoot,
		setCurrentStateRoot: (root) => {
			if (!state.stateRoots.has(root)) {
				throw new Error('Cannot set state root to non existing state root')
			}
			currentStateRoot = root
		},
		stateRoots,
		options,
		caches: {
			contracts: new ContractCache(),
			accounts: new AccountCache({
				size: 100_000,
				type: CacheType.ORDERED_MAP,
			}),
			storage: new StorageCache({
				size: 100_000,
				type: CacheType.ORDERED_MAP,
			}),
		},
		ready: () => genesisPromise.then(() => true),
	}
	const genesisPromise = (
		options.genesisState !== undefined && options.currentStateRoot === undefined
			? generateCanonicalGenesis(state)(options.genesisState)
			: Promise.resolve().then(() => {
					if (options.currentStateRoot) {
						state.setCurrentStateRoot(options.currentStateRoot)
						return generateCanonicalGenesis(state)(options.genesisState ?? stateRoots.get(options.currentStateRoot))
					}
					return Promise.resolve()
				})
	).then(async () => {
		await checkpoint(state)()
		await commit(state)()
		logger.debug('StateManager is ready')
	})

	return state
}
