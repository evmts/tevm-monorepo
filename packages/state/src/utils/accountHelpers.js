import { Rlp } from '@tevm/rlp'
import { createAccount, createAddressFromString as createAddress } from '@tevm/utils'

/**
 * Helper to create an address from a string
 * @param {string} addressString - The address string
 * @returns {import('@tevm/utils').EthjsAddress} The address
 */
export function createAddressFromString(addressString) {
	return createAddress(addressString)
}

/**
 * Creates an account from RLP serialized data
 * @param {Uint8Array} serialized - The RLP serialized account data
 * @returns {import('@tevm/utils').EthjsAccount} The account
 */
export function fromRlpSerializedAccount(serialized) {
	const decoded = Rlp.decode(serialized)
	if (!Array.isArray(decoded) || decoded.length !== 4) {
		throw new Error('Invalid RLP serialized account')
	}

	const [nonce, balance, storageRoot, codeHash] = decoded

	/** @type {{nonce: any, balance: any, storageRoot: any, codeHash: any}} */
	const accountData = {
		nonce,
		balance,
		storageRoot,
		codeHash,
	}

	return createAccount(accountData)
}

/**
 * Creates an account from account data
 * @param {{nonce?: bigint, balance?: bigint, storageRoot?: Uint8Array, codeHash?: Uint8Array}} accountData - The account data
 * @returns {import('@tevm/utils').EthjsAccount} The account
 */
export function fromAccountData(accountData) {
	return createAccount(accountData)
}
