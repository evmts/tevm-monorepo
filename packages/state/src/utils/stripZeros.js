import { InternalError } from '@tevm/errors'

/**
 * @param {Uint8Array} bytes
 * @returns {Uint8Array}
 */
export const stripZeros = (bytes) => {
	if (!(bytes instanceof Uint8Array)) {
		throw new InternalError('Unexpected type')
	}
	return bytes.slice(bytes.findIndex(/** @param {number} entry*/ (entry) => entry !== 0))
}
