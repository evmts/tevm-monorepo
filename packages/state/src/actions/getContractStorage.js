import { bytesToHex, hexToBytes } from 'viem'
import { getForkBlockTag } from './getForkBlockTag.js'
import { getForkClient } from './getForkClient.js'
import { putContractStorage } from './putContractStorage.js'

/**
 * Gets the storage value associated with the provided `address` and `key`. This method returns
 * the shortest representation of the stored value.
 * corresponding to the provided address at the provided key.
 * If this does not exist an empty `Uint8Array` is returned.
 * @type {import("../state-types/index.js").StateAction<'getContractStorage'>}
 */
export const getContractStorage = (baseState) => async (address, key) => {
	const {
		caches: { storage: storageCache },
	} = baseState
	// Check storage slot in cache
	if (key.length !== 32) {
		throw new Error('Storage key must be 32 bytes long')
	}

	const cachedValue = storageCache.get(address, key)
	if (cachedValue !== undefined) {
		return cachedValue
	}

	if (!baseState.options.fork?.url) {
		return hexToBytes('0x0')
	}

	baseState.logger.debug({ address, key }, 'Fetching storage from remote RPC')

	const client = getForkClient(baseState)
	const blockTag = getForkBlockTag(baseState)

	const storage = await client.getStorageAt({
		address: /** @type {import('@tevm/utils').Address} */ (address.toString()),
		slot: bytesToHex(key),
		...blockTag,
	})
	const value = hexToBytes(storage ?? '0x0')

	await putContractStorage(baseState)(address, key, value)

	baseState.logger.debug({ address, key, value }, 'Cached forked storage to state')

	return value
}
