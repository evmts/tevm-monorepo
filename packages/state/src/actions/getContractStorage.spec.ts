import { createAddress } from '@tevm/address'
import { transports } from '@tevm/test-utils'
import { EthjsAccount, EthjsAddress, hexToBytes, toBytes } from '@tevm/utils'
import { beforeEach, describe, expect, it } from 'vitest'
import type { BaseState } from '../BaseState.js'
import { createBaseState } from '../createBaseState.js'
import { getContractStorage } from './getContractStorage.js'
import { putAccount } from './putAccount.js'
import { putContractStorage } from './putContractStorage.js'

describe('getContractStorage', () => {
	let baseState: BaseState
	let address: EthjsAddress
	let key: Uint8Array
	let value: Uint8Array
	let account: EthjsAccount

	beforeEach(async () => {
		baseState = createBaseState({
			loggingLevel: 'warn',
		})

		address = createAddress('01'.repeat(20))
		key = hexToBytes(`0x${'02'.repeat(32)}`)
		value = hexToBytes('0x1234')
		account = EthjsAccount.fromAccountData({
			balance: 420n,
			nonce: 2n,
		})

		await putAccount(baseState)(address, account)
		await putContractStorage(baseState)(address, key, value)
	})

	it('should get the storage value associated with the provided address and key', async () => {
		expect(await getContractStorage(baseState)(address, key)).toEqual(value)
	})

	it('should return empty Uint8Array if the storage does not exist', async () => {
		const newKey = hexToBytes(`0x${'03'.repeat(32)}`)
		expect(await getContractStorage(baseState)(address, newKey)).toEqual(Uint8Array.from([0]))
	})

	it('should throw an error if the key is not 32 bytes long', async () => {
		const invalidKey = hexToBytes('0x1234')
		const err = await getContractStorage(baseState)(address, invalidKey).catch((e) => e)
		expect(err).toBeInstanceOf(Error)
		expect(err.message).toMatchSnapshot()
	})

	it('should return empty Uint8Array if the account is not a contract', async () => {
		const newAddress = EthjsAddress.fromString(`0x${'02'.repeat(20)}`)
		await putAccount(baseState)(newAddress, EthjsAccount.fromAccountData({ balance: 100n, nonce: 1n }))
		expect(await getContractStorage(baseState)(newAddress, key)).toEqual(Uint8Array.from([0]))
	})
})

describe('getContractStorage forking', () => {
	let baseState: BaseState
	let knownContractAddress: EthjsAddress
	let knownStorageKey: Uint8Array

	beforeEach(() => {
		baseState = createBaseState({
			loggingLevel: 'warn',
			fork: {
				transport: transports.optimism,
				blockTag: 122488188n,
			},
		})

		knownContractAddress = EthjsAddress.fromString('0x4200000000000000000000000000000000000042')
		knownStorageKey = toBytes(1, { size: 32 })
	})

	it('should fetch storage from remote provider if not in cache and fork transport is provided', async () => {
		const result = await getContractStorage(baseState)(knownContractAddress, knownStorageKey)
		expect(result).toBeDefined()
		expect(result.length).toBeGreaterThan(0)
		expect(result).toMatchSnapshot()
		const cachedResult = await getContractStorage(baseState)(knownContractAddress, knownStorageKey)
		expect(cachedResult).toEqual(result)
	})

	it('should return empty Uint8Array if the account does not exist and no fork transport', async () => {
		const noForkBaseState = createBaseState({
			loggingLevel: 'warn',
		})
		expect(await getContractStorage(noForkBaseState)(knownContractAddress, knownStorageKey)).toMatchInlineSnapshot(`
			Uint8Array [
			  0,
			]
		`)
	})
})
