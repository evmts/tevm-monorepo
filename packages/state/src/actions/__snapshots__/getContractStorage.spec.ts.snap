// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getContractStorage > should throw an error if the key is not 32 bytes long 1`] = `
"Storage key must be 32 bytes long. Received 2. If using numberToHex make the length 32.

Docs: https://tevm.sh/reference/tevm/errors/classes/internalerror/
Version: 1.1.0.next-73"
`;

exports[`getContractStorage forking > should fetch storage from remote provider if not in cache and fork transport is provided 1`] = `
Uint8Array [
  0,
]
`;
